cartographer/mapping/internal/3d/pose_graph_3d.cc

```diff
@@ -152,26 +153,19 @@ NodeId PoseGraph3D::AddNode(
 
 void PoseGraph3D::AddWorkItem(
     const std::function<WorkItem::Result()>& work_item) {
-  {
-    common::MutexLocker locker(&work_queue_mutex_);
-    if (work_queue_ != nullptr) {
-      const auto now = std::chrono::steady_clock::now();
-      work_queue_->push_back({now, work_item});
-      kWorkQueueDelayMetric->Set(
-          common::ToSeconds(now - work_queue_->front().time));
-      return;
-    }
-  }
-  if (work_item() == WorkItem::Result::kRunOptimization) {
-    {
-      common::MutexLocker locker(&work_queue_mutex_);
-      work_queue_ = common::make_unique<WorkQueue>();
-    }
-    constraint_builder_.WhenDone(
-        [this](const constraints::ConstraintBuilder3D::Result& result) {
-          HandleWorkQueue(result);
-        });
+  common::MutexLocker locker(&work_queue_mutex_);
+  if (work_queue_ == nullptr) {
+    work_queue_ = common::make_unique<WorkQueue>();
+    auto task = common::make_unique<common::Task>();
+    task->SetWorkItem([this]() { DrainWorkQueue(); });
+    thread_pool_->Schedule(std::move(task));
   }
+  const auto now = std::chrono::steady_clock::now();
+  work_queue_->push_back({now, work_item});
+  kWorkQueueDelayMetric->Set(
```

Comment:
compared to before we now call `Set` in both cases `work_queue_ == nullptr`  and `work_queue_ != nullptr`, but for me this change looks still fine.

Reply:
Yes, because before we would basically say that when the work queue was idle and we run a single task immediately in the foreground, that this does not constitute a delay.